src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs

```diff
@@ -128,6 +134,63 @@ public override void Visit(IEnumerable<ITypeDefinitionMember> members)
             base.Visit(members);
         }
 
+        public override void Visit(ITypeDefinition parentType, IEnumerable<IFieldDefinition> fields)
+        {
+            if (parentType.IsStruct && !_apiOnly)
+            {
+                // For compile-time compat, the following rules should work for producing a reference assembly. We drop all private fields, but add back certain synthesized private fields for a value type (struct) as follows:
+                // - If there are any private fields that are or contain any value type members, add a single private field of type int
+                // - If there are any private fields that are or contain any reference type members, add a single private field of type object.
+                // - If the type is generic, then for every type parameter of the type, if there are any private fields that are or contain any members whose type is that type parameter, we add a direct private field of that type.
+
```

Comment:
"contain any members whose type is that type parameter"   -  I do not understand how this part is handled. 

Are the input "fields" actually a flattened recursive list of all instance fields and fields of struct fields?
Or is there is some other way this recurses into fields of fields?

Do we actually have a case where fields of generic types are contained indirectly.
Like:

```C#
struct ContainsFieldOfT<T>
{
      // no fields of type T contained directly, but there is one nested 1 level deeper
      private Nullable<T> _field;
}

Reply:
I see. Makes sense. Thanks!

---
In reply to: [160841720](https://github.com/dotnet/buildtools/pull/1859#discussion_r160841720) [](ancestors = 160841720)